networkx/algorithms/approximation/steinertree.py

```diff
@@ -46,22 +57,172 @@ def metric_closure(G, weight="weight"):
     return M
 
 
+def mehlhorn_steiner_tree(G, S, weight):
+    paths = nx.multi_source_dijkstra_path(G, S)
+
+    d_1 = {}
+    s = {}
+    for v in G.nodes():
+        s[v] = paths[v][0]
+        d_1[(v, s[v])] = len(paths[v]) - 1
+
+    new_edges = {}
+    for e in G.edges(data=True):
+        u, v, data = e
+        weight_here = d_1[(u, s[u])] + data.get(weight, 1) + d_1[(v, s[v])]
+        key = tuple(sorted([s[u], s[v]]))
```

Comment:
In general, the nodes are not sortable, so you should find a way to avoid the sort on `key`. You can use a graph data structure to help with this.  Set the edge on G_1_prime and then check it the next time. The edge weight is connected to both (u, v) and (v, u) so the order doesn't matter.  Something like:

```python
    G_1_prime = nx.Graph()
    for u, v, data  in G.edges(data=True):
        su, sv = s[u], s[v]
        weight_here = d_1[(u, su)] + data.get(weight, 1) + d_1[(v, sv)]
        if not G_1_prime.has_edge(su, sv):
            G_1_prime.add_edge(su, sv, weight=weight_here)
        else:
            new_weight = min(weight_here, G_1_prime[su][sv][weight])
            G_1_prime.add_edge(su, sv, weight=new_weight)
```